### PR TITLE
Remove the assertion to unblock breakages

### PR DESCRIPTION
**Pull request by wlei-llvm**
as titled.

### CODE REVIEW

**Review by davidtgoldblatt — APPROVED** [review]
Looks good in general; 1 nit before merging: should be "redfined", not "derefined".

**Comment by wlei-llvm** [comment]
> Looks good in general; 1 nit before merging: should be "redfined", not "derefined".

I see. Here both refinement and de-refinement can lead to a different version of IR (the checksum mismatch) in link time. How about just use "NonExactDefinition"? 
(I thought the opposite of the `isDefinitionExact` was `Derefined` while I was looking at the code..)
```
  bool isDefinitionExact() const {
    return !mayBeDerefined();
  }
```

**Comment by wlei-llvm** [comment]
Hmm, tested on our internal services, it still ran into the assertion.. the function linkage type is `internal`, it's very weird.

**Comment by wlei-llvm** [comment]
> Hmm, tested on our internal services, it still ran into the assertion.. the function linkage type is `internal`, it's very weird. How would an `internal` function be replaced by another definition.

OK, so this is because there is no profile loaded for this function in pre-link time (so it never run this `profileIsValid`), the attr is not set(unknown). 
but in post-link time, profile is loaded(likely from other func's children profile), then find the state is mismatch.

This is probably not easy to quickly fix.. (Seems we need  a three state for the attr: matched, mismatch, unknown..., maybe we don't need those complexity or just remove the assertion) 

cc @WenleiHe

**Comment by wlei-llvm** [comment]
Let me remove the assertion to unblock the build, will dig more later.

**Comment by WenleiHe** [comment]
> OK, so this is because there is no profile loaded for this function in pre-link time (so it never run this `profileIsValid`), the attr is not set(unknown).

Sounds good to remove assertion to unblock.

> OK, so this is because there is no profile loaded for this function in pre-link time (so it never run this profileIsValid), the attr is not set(unknown).

I'm curious why does this happen?

**Comment by wlei-llvm** [comment]
> > OK, so this is because there is no profile loaded for this function in pre-link time (so it never run this profileIsValid), the attr is not set(unknown).
> 
> I'm curious why does this happen?

This can happen when meet:
1) there is no outline/top-level function in the profile( the function is all inlined into its caller or pre-inliner all inlines them). 
2) because we use flattened profile. In prelink, all the internal function's nested(children) profile doesn't have this function.
3) In post-link, one (external) function 's nested profile have this function. 

basically this is caused by the profile flattening.

**Comment by ayermolo** [comment]
> > > OK, so this is because there is no profile loaded for this function in pre-link time (so it never run this profileIsValid), the attr is not set(unknown).
> > 
> > 
> > I'm curious why does this happen?
> 
> This can happen when meet:
> 
> 1. there is no outline/top-level function in the profile( the function is all inlined into its caller or pre-inliner all inlines them).
> 2. because we use flattened profile. In prelink, all the internal function's nested(children) profile doesn't have this function.
> 3. In post-link, one (external) function 's nested profile have this function.
> 
> basically this is caused by the profile flattening.

What should happen in that case?
Is the test warranted?

**Comment by wlei-llvm** [comment]
> 
> What should happen in that case? Is the test warranted?

Before we expect a consistency(the assertion) between the mismatch state computed in pre-link and attr used in post-link, but because of this "issue", we understand that  attr is not always equal to the pre-link state. Now I think we may just live with this, for local symbol, just use the original state, attr is only used when original state is missing(imported function). 

Sorry for my bad linker knowledge, I thought the inconsistency was from the replacement of definition.

**Comment by WenleiHe** [comment]
> > > OK, so this is because there is no profile loaded for this function in pre-link time (so it never run this profileIsValid), the attr is not set(unknown).
> > 
> > 
> > I'm curious why does this happen?
> 
> This can happen when meet:
> 
> 1. there is no outline/top-level function in the profile( the function is all inlined into its caller or pre-inliner all inlines them).
> 2. because we use flattened profile. In prelink, all the internal function's nested(children) profile doesn't have this function.
> 3. In post-link, one (external) function 's nested profile have this function.
> 
> basically this is caused by the profile flattening.

Ok, so the function's profile only exists as nested inlinee profile in a different module, so for pre-link compilation of that module, we don't have a profile to check for mismatch.

File: llvm/include/llvm/Transforms/Utils/SampleProfileLoaderBaseImpl.h
Comment (WenleiHe) [inline llvm/include/llvm/Transforms/Utils/SampleProfileLoaderBaseImpl.h:142]: nit: leave a TODO comment?

**Review by WenleiHe — APPROVED** [review]
LGTM to unblock, thanks.

**Pull request merged**
